src/common/expression/src/evaluator.rs

```diff
@@ -108,13 +118,18 @@ impl<'a> Evaluator<'a> {
 
         #[cfg(debug_assertions)]
         if result.is_err() {
-            assert_eq!(
-                ConstantFolder::new(&self.input_columns.domains())
-                    .fold(expr)
-                    .1,
-                None,
-                "domain calculation should not return any domain for expressions that are possible to fail"
-            );
+            static RECURSING: Mutex<bool> = Mutex::new(false);
```

Comment:
`use parking_lot::Mutex;`
No need `unwarp` here

Reply:
This is for debugging only, so I think it won't bother adding an extra dependency.